### ISSUE TITLE
optimize selection process?

### ISSUE THREAD

**Issue by nickhand** [bccp/nbodykit#391]
I ran into an annoying use case where an input catalog included invalid data (in this case, negative redshifts). I sliced the object to include only valid data, but only after doing my data cleaning. In this case, I added "Position" (which cannot be computed for z<0). When computing Position, this fails because the selection is the last step performed (the dask engine does all of the operations on the whole catalog first).  You can say this is a user error (I should have sliced first) but I would argue it's likely a common mistake and one that would be nice to optimize away.  

In theory, this can be optimized away for potentially significant gains. The size of the catalog is fixed after some point in the dask graph, so we can insert the selection tasks into the dask graph after the size is finalized. 

I am working on a potential implementation of this -- it delves into the guts of dask a bit so we'll need to see if it is worth it.

**Comment by rainwoodman** [comment]
I still don't quite understand this. could you post a code segment?

**Comment by nickhand** [comment]
Currently this fails:

```python
s = RandomCatalog(1000, seed=42)
s['z']   = s.rng.normal(loc=0, scale=0.2, size=s.size) # contains z < 0
s['ra']  = s.rng.uniform(low=110, high=260, size=s.size)
s['dec'] = s.rng.uniform(low=-3.6, high=60., size=s.size)

# add Position
s['Position'] = transform.SkyToCartesion(s['ra'], s['dec'], s['z'], cosmo=cosmology.Planck15)

# select valid range
subset = s[s['z'] > 0]

# fails with CLASS error for z<0 
pos = subset['Position'].compute()
```

**Comment by rainwoodman** [comment]
I see -- I was confused with the selection column. The selection is performed after compute. I'll take a better look at 392 tomorrow.

**Comment by nickhand** [comment]
closed via #392 included as part of #393
